packages/app-builder/src/components/Data/DataModelFlow.tsx

```diff
@@ -0,0 +1,269 @@
+  DataModelNodeData,
+  DataModelEdgeData
+>;
+
+interface DataModelFlowProps {
+  dataModel: TableModel[];
+  children?: React.ReactNode;
+}
+
+export const dataModelFlowStyles = reactflowStyles;
+
+export function DataModelFlow({ dataModel, children }: DataModelFlowProps) {
+  return (
+    <ReactFlowProvider>
+      <DataModelFlowImpl dataModel={dataModel}>{children}</DataModelFlowImpl>
+    </ReactFlowProvider>
+  );
+}
+
+function DataModelFlowImpl({ dataModel, children }: DataModelFlowProps) {
+  const { t } = useTranslation(dataI18n);
+  const [nodes, setNodes] = React.useState<Array<Node<DataModelNodeData>>>([]);
+  const [edges, setEdges] = React.useState<Array<Edge<DataModelEdgeData>>>([]);
+
+  const onNodesChange = React.useCallback((changes: NodeChange[]) => {
+    const allowedChanges = changes.filter((change) => change.type !== 'remove');
+    setNodes((nds) => applyNodeChanges(allowedChanges, nds));
+  }, []);
+  const onEdgesChange = React.useCallback((changes: EdgeChange[]) => {
+    const allowedChanges = changes.filter((change) => change.type !== 'remove');
+    setEdges((eds) => applyEdgeChanges(allowedChanges, eds));
+  }, []);
+
+  // Update nodes and edges when dataModel changes
+  React.useEffect(() => {
+    setNodes((currentNodes) =>
+      R.pipe(
+        dataModel,
+        R.map((tableModel) => adaptTableModelNode(tableModel, dataModel)),
+        R.map((newNode) => {
+          const existingNode = currentNodes.find((nd) => nd.id === newNode.id);
+          if (!existingNode) {
+            newNode.style = { ...newNode.style, opacity: 0 };
+            return newNode;
+          }
+          // Only update data
+          return { ...existingNode, data: newNode.data };
+        }),
+      ),
+    );
+    setEdges((currentEdges) =>
+      R.pipe(
+        dataModel,
+        R.flatMap((tableModel) =>
+          tableModel.linksToSingle.map((link) =>
+            adaptDataModelEdges(tableModel, link),
+          ),
+        ),
+        R.map((newEdge) => {
+          const existingEdge = currentEdges.find((ed) => ed.id === newEdge.id);
+          if (!existingEdge) {
+            newEdge.style = { ...newEdge.style, opacity: 0 };
+            return newEdge;
+          }
+          // Only update data
+          return { ...existingEdge, data: newEdge.data };
+        }),
+      ),
+    );
+  }, [dataModel]);
+
+  const { fitView } = useDataModelReactFlow();
+  React.useEffect(() => {
+    // Wait first render of each node to have dynamic width before layouting
+    if (nodes.some((nd) => nd.width === undefined)) return;
+
+    if (
+      nodes.some((nd) => nd.style?.opacity === 0) ||
+      edges.some((ed) => ed.style?.opacity === 0)
+    ) {
+      const layouted = getLayoutedElements(nodes, edges);
+      setNodes(layouted.nodes);
+      setEdges(layouted.edges);
+      window.requestAnimationFrame(() => {
+        fitView();
+      });
+    }
+  }, [fitView, edges, nodes]);
```

Comment:
This is where the reconciliation logic + opacity hack is implemented.
The rest is just "basic code"